runtime/gc_modron_startup/mminit.cpp

```diff
@@ -2907,7 +2908,10 @@ gcInitializeDefaults(J9JavaVM* vm)
 
 				if (hwSupported) {
 					/* Software Barrier request overwrites HW usage on supported HW */
-					extensions->concurrentScavengerHWSupport = hwSupported && !extensions->softwareRangeCheckReadBarrier && !J9_ARE_ANY_BITS_SET(vm->extendedRuntimeFlags2, J9_EXTENDED_RUNTIME2_ENABLE_PORTABLE_SHARED_CACHE);
+					extensions->concurrentScavengerHWSupport = hwSupported 
+						&& !extensions->softwareRangeCheckReadBarrier 
+						&& !J9_ARE_ANY_BITS_SET(vm->extendedRuntimeFlags2, J9_EXTENDED_RUNTIME2_ENABLE_PORTABLE_SHARED_CACHE)
+						&& !J9_ARE_ANY_BITS_SET(vm->extendedRuntimeFlags2, J9_EXTENDED_RUNTIME2_ENABLE_CRIU_SUPPORT);
```

Comment:
Maybe this is still needed? Unless there's going to be some other changes in the future to catch environment differences like this?

Reply:
Am I reading this correctly "Restored (by CRIU) image can not support Guarded Storage HW" ?

Reply:
Initially I blindly just grepped for `J9_EXTENDED_RUNTIME2_ENABLE_PORTABLE_SHARED_CACHE` and added `J9_EXTENDED_RUNTIME2_ENABLE_CRIU_SUPPORT` as well. After https://github.com/eclipse-openj9/openj9/pull/13113#issuecomment-875674737 I revisited this PR and I believe the compressed refs shift shouldn't be fixed (comment above), but I don't know about HW GS; we'd have to ensure that the restored JVM runs on a machine that also has HW support no?

Reply:
Yes, we need either control both saved/restored runs on the same type of HW or just disable usage of GS in case image is going to be saved or have restored (if we can afford)

Reply:
Using CRIU split execution for two cases obviously (due nature of saving/restoring process image):
- JVM startup/execution ends at image creation
- JVM has been restored from image
Is `J9_EXTENDED_RUNTIME2_ENABLE_CRIU_SUPPORT` flag intended to recognize first case only? Or, alternatively, it might be used for both but it creates some uncertainty.

Reply:
Heh, the answer to that depends on what happens in https://github.com/eclipse-openj9/openj9/pull/13113. At the moment the flag can't distinguish between the snapshot or the restore run. https://github.com/eclipse-openj9/openj9/pull/13113 may add another flag that can distinguish between the two.

That said, can we run in mixed mode GS? I.e. SW read barriers pre snapshot, and then HW read barriers (for new JIT code) post restore (even ignoring all of the initialization that has to happen on every single Java Thread).

Reply:
#13113 has been merged. The major changes are that you need to use `vmFuncs->isCRIUSupportEnabled` to determine if criu support is enabled (-XX:+EnableCRIUSupport). `vmFuncs->isCheckpointAllowed` is used to determine if the VM is capable of checkpointing. If `-XX:+CRIURestoreNonPortableMode` is supplied, `isCheckpointAllowed` will return FALSE after the first checkpoint